### ISSUE TITLE
Enum implicit cast to string fails when element is named "ToString"

### ISSUE THREAD

**Issue by scalablecory** [dotnet/roslyn#40256]
_From @agustins on Tuesday, November 19, 2019 10:05:16 PM_

When you have an enum with an element named "ToString" compiler fails with

`"csc.exe" exited with code -2146232797. ...Bin\Roslyn\Microsoft.CSharp.Core.targets	59`

# Environment
 * .NET Core 3.0 and SDK Version 3.0.100
 * Visual Studio Enterprise 2019 Version 16.3.9

# Reproduce

``` csharp
using System;

namespace ConsoleApp
{
    class Program
    {
        public enum Foo
        {
            ToString
        }

        static void Main(string[] args)
        {
            Console.WriteLine(Foo.ToString + "");
        }
    }
}
```

_Copied from original issue: dotnet/core#3850_

**Comment by scalablecory** [comment]
_From @carlossanlop on Tuesday, November 19, 2019 10:48:32 PM_

I cannot repro: I was able to build and run the app with no issues. This is a problem with something else, problably with VS or your .NET installation. If you rename the enum value to something else, and compile again, are you getting the same error? Can you paste the full error message?

This exact error number has been reported in the past in other places. For example [here](https://stackoverflow.com/questions/51168359/csc-exe-exited-with-code-2146232797-vs-2017), which ended up being a problem with VS 2017. But you have 2019, so I'd like to know the full error, if you can share it.

**Comment by scalablecory** [comment]
_From @agustins on Wednesday, November 20, 2019 5:16:51 PM_

If I rename the value to "ToString2" for example builds successfully.

The exact error:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB6006	"csc.exe" exited with code -2146232797.	ConsoleApp2	C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Current\Bin\Roslyn\Microsoft.CSharp.Core.targets	59	
```


csc version: 3.300.119.46102

**Comment by scalablecory** [comment]
_From @SusLes on Thursday, December 5, 2019 9:16:24 PM_

@carlossanlop

**Comment by yaakov-h** [comment]
That’s the exit code for COR_E_FAILFAST. That’ll probably need a debugger to uncover the actual error.

**Comment by svick** [comment]
@yaakov-h [Or you can paste the code into SharpLab, which shows that the exception is](https://sharplab.io/#v2:D4AQTAjAsAULB2BDAtgUwM4AdEGNUAIBhAe3nWIBtUBBTTWAb1nxf3DYgHZnWmZWBbAMz5U8AK7J8AMWLEeg/H0WKAKsQDKAFwBOAS3gBzBYIC+sEwJAQAbGwAs+ALKIDACmsAGANoBdfIg6hugAlJa84YrWAJxussQAdOra+kb4ANT4AERZIQDckfjm/KzFpkA=):

```
System.InvalidCastException: Unable to cast object of type 'ZeroValuedEnumConstantSymbol' to type 'Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol'.
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.ConvertConcatExprToString(SyntaxNode syntax, BoundExpression expr)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.RewriteStringConcatenation(SyntaxNode syntax, BinaryOperatorKind operatorKind, BoundExpression loweredLeft, BoundExpression loweredRight, TypeSymbol type)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.MakeBinaryOperator(BoundBinaryOperator oldNode, SyntaxNode syntax, BinaryOperatorKind operatorKind, BoundExpression loweredLeft, BoundExpression loweredRight, TypeSymbol type, MethodSymbol method, Boolean isPointerElementAccess, Boolean isCompoundAssignment, BoundUnaryOperator applyParentUnaryOperator)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitBinaryOperator(BoundBinaryOperator node, BoundUnaryOperator applyParentUnaryOperator)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitBinaryOperator(BoundBinaryOperator node)
   at Microsoft.CodeAnalysis.CSharp.BoundBinaryOperator.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionImpl(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriter.DoVisitList[T](ImmutableArray`1 list)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitCall(BoundCall node)
   at Microsoft.CodeAnalysis.CSharp.BoundCall.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeRewriterWithStackGuard.VisitExpressionWithoutStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.BoundTreeVisitor.VisitExpressionWithStackGuard(Int32& recursionDepth, BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionImpl(BoundExpression node)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitUnusedExpression(BoundExpression expression)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.RewriteExpressionStatement(BoundExpressionStatement node, Boolean suppressInstrumentation)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitExpressionStatement(BoundExpressionStatement node)
   at Microsoft.CodeAnalysis.CSharp.BoundExpressionStatement.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitPossibleUsingDeclaration(BoundStatement node, ImmutableArray`1 statements, Int32 statementIndex, Boolean& replacedLocalDeclarations)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitStatementSubList(ArrayBuilder`1 builder, ImmutableArray`1 statements, Int32 startIndex)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.VisitBlock(BoundBlock node)
   at Microsoft.CodeAnalysis.CSharp.BoundBlock.Accept(BoundTreeVisitor visitor)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Visit(BoundNode node)
   at Microsoft.CodeAnalysis.CSharp.LocalRewriter.Rewrite(CSharpCompilation compilation, MethodSymbol method, Int32 methodOrdinal, NamedTypeSymbol containingType, BoundStatement statement, TypeCompilationState compilationState, SynthesizedSubmissionFields previousSubmissionFields, Boolean allowOmissionOfConditionalCalls, Boolean instrumentForDynamicAnalysis, ImmutableArray`1& dynamicAnalysisSpans, DebugDocumentProvider debugDocumentProvider, DiagnosticBag diagnostics, Boolean& sawLambdas, Boolean& sawLocalFunctions, Boolean& sawAwaitInExceptionHandler)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.LowerBodyOrInitializer(MethodSymbol method, Int32 methodOrdinal, BoundStatement body, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState, Boolean instrumentForDynamicAnalysis, DebugDocumentProvider debugDocumentProvider, ImmutableArray`1& dynamicAnalysisSpans, DiagnosticBag diagnostics, VariableSlotAllocator& lazyVariableSlotAllocator, ArrayBuilder`1 lambdaDebugInfoBuilder, ArrayBuilder`1 closureDebugInfoBuilder, StateMachineTypeSymbol& stateMachineTypeOpt)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethod(MethodSymbol methodSymbol, Int32 methodOrdinal, ProcessedFieldInitializers& processedInitializers, SynthesizedSubmissionFields previousSubmissionFields, TypeCompilationState compilationState)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileNamedType(NamedTypeSymbol containingType)
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.<>c__DisplayClass22_0.<CompileNamedTypeAsTask>b__0()
   at Roslyn.Utilities.UICultureUtilities.<>c__DisplayClass5_0.<WithCurrentUICulture>b__0()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.WaitForWorkers()
   at Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(CSharpCompilation compilation, PEModuleBuilder moduleBeingBuiltOpt, Boolean emittingPdb, Boolean emitTestCoverageData, Boolean hasDeclarationErrors, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.CompileMethods(CommonPEModuleBuilder moduleBuilder, Boolean emittingPdb, Boolean emitMetadataOnly, Boolean emitTestCoverageData, DiagnosticBag diagnostics, Predicate`1 filterOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream metadataPEStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, CompilationTestData testData, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Compilation.Emit(Stream peStream, Stream pdbStream, Stream xmlDocumentationStream, Stream win32Resources, IEnumerable`1 manifestResources, EmitOptions options, IMethodSymbol debugEntryPoint, Stream sourceLinkStream, IEnumerable`1 embeddedTexts, Stream metadataPEStream, CancellationToken cancellationToken)
```

**Comment by yaakov-h** [comment]
@svick yes, or you could do that, much easier 😄

**Comment by jcouv** [comment]
This is a known issue (issue https://github.com/dotnet/roslyn/issues/38858). Was fixed in 16.4 preview 2 (PR https://github.com/dotnet/roslyn/issues/38858).
I'll go ahead and close the issue as duplidate. Thanks
